pkg/providers/vmware/provider.go

```diff
@@ -0,0 +1,422 @@
+		return err
+	}
+	r.vm = vm.(*object.VirtualMachine)
+	return nil
+}
+
+// GetVMName gets the name of the source VM
+func (r *VmwareProvider) GetVMName() (string, error) {
+	vm, err := r.getVM()
+	if err != nil {
+		return "", err
+	}
+	return vm.Name(), nil
+}
+
+// GetVMStatus gets the power status of the source VM
+func (r *VmwareProvider) GetVMStatus() (provider.VMStatus, error) {
+	vmProperties, err := r.getVmProperties()
+	if err != nil {
+		return "", err
+	}
+
+	poweredOn := types.VirtualMachinePowerStatePoweredOn
+	poweredOff := types.VirtualMachinePowerStatePoweredOff
+
+	switch vmProperties.Runtime.PowerState {
+	case poweredOn:
+		return provider.VMStatusUp, nil
+	case poweredOff:
+		return provider.VMStatusDown, nil
+	}
+
+	return "", fmt.Errorf("VM doesn't have a legal status. Allowed statuses: [%v, %v]", poweredOn, poweredOff)
+}
+
+// StartVM powers on the source VM.
+func (r *VmwareProvider) StartVM() error {
+	vmwareClient, err := r.getClient()
+	if err != nil {
+		return err
+	}
+	vm, err := r.getVM()
+	if err != nil {
+		return err
+	}
+	return vmwareClient.StartVM(vm.Reference().Value)
+}
+
+// StartVM powers off the source VM.
+func (r *VmwareProvider) StopVM(cr *v1beta1.VirtualMachineImport, client client.Client) error {
+	vmwareClient, err := r.getClient()
+	if err != nil {
+		return err
+	}
+	vm, err := r.getVM()
+	if err != nil {
+		return err
+	}
+	err = vmwareClient.StopVM(vm.Reference().Value)
+	if err != nil {
+		return err
+	}
+	err = utils.AddFinalizer(cr, utils.RestoreVMStateFinalizer, client)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CleanUp removes transient resources created for import
+func (r *VmwareProvider) CleanUp(failure bool, cr *v1beta1.VirtualMachineImport, client client.Client) error {
+	var errs []error
+
+	err := utils.RemoveFinalizer(cr, utils.RestoreVMStateFinalizer, client)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	vmiName := k8stypes.NamespacedName{
+		Name:      r.vmiObjectMeta.Name,
+		Namespace: r.vmiObjectMeta.Namespace,
+	}
+
+	err = r.secretsManager.DeleteFor(vmiName)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	err = r.configMapsManager.DeleteFor(vmiName)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	if failure {
+		err = r.dataVolumesManager.DeleteFor(vmiName)
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		err = r.virtualMachineManager.DeleteFor(vmiName)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		return foldErrors(errs, vmiName)
+	}
+	return nil
+}
+
+// TestConnection tests the connection to the vCenter or ESXi host.
+func (r *VmwareProvider) TestConnection() error {
+	vmwareClient, err := r.getClient()
+	if err != nil {
+		return err
+	}
+	return vmwareClient.TestConnection()
+}
+
+// Validate checks whether the source VM and resource mapping is valid.
+func (r *VmwareProvider) Validate() ([]v1beta1.VirtualMachineImportCondition, error) {
+	// TODO: implement vmware rule validation
```

Comment:
Will both source VM compatibility validation and resource mapping be implemented in the external service?

Reply:
@fdupont-redhat do you know whether the shared validation service will handle mappings?

Reply:
@mansam probably not in MVP, but in my opinion it should handle both. The inventory service will collect both VMware and Kubevirt data, so it will be able to request a validation for the mapping.